src/java.base/share/classes/java/util/concurrent/locks/StampedLock.java

```diff
@@ -1244,21 +1256,25 @@ else if (!casTail(t, node))
                 if (node.waiter == null)
                     node.waiter = Thread.currentThread();
                 node.status = WAITING;
-            } else {
-                long nanos;
-                spins = postSpins = (byte)((postSpins << 1) | 1);
-                if (!timed)
-                    LockSupport.park(this);
-                else if ((nanos = time - System.nanoTime()) > 0L)
-                    LockSupport.parkNanos(this, nanos);
-                else
-                    break;
+            } else if (!timed || (nanos = time - System.nanoTime()) > 0L) {
+                try {
+                    if (!timed)
+                        LockSupport.park(this);
+                    else
+                        LockSupport.parkNanos(this, nanos);
+                } catch (Error | RuntimeException ex) {
+                    cancelAcquire(node);
+                    throw ex;
+                }
                 node.clearStatus();
                 if ((interrupted |= Thread.interrupted()) && interruptible)
                     break;
-            }
+                spins = postSpins = (byte)((postSpins << 1) | 1);
```

Comment:
@DougLea Any specific reason to move the updates to `spins` and `postSpins` to after the park?

Reply:
Just to improve clarity after the exceptional and break cases

Reply:
Makes sense—thanks